### PR TITLE
feat: migrate to ape and add features

### LINKED ISSUE #15: rewrite to ape and update
assigned to @milkyklim

- [x] rewrite from brownie to ape
- [x] upgrade to latest vyper
- [x] add tests for newly merged features
- [ ] clean up contract
- [ ] identify other features to add/remove
- [x] use new vyper blueprints in the factory
- [ ] get a review from @banteg @fubuloubu and @storming0x
- [ ] deploy the new version

### PR DESCRIPTION
**Pull request by milkyklim**
closes #15

### CODE REVIEW

File: contracts/VestingEscrowSimple.vy [hunk 1]
```diff
@@ -70,23 +62,24 @@ def initialize(
...
     self.recipient = recipient
     self.disabled_at = end_time  # Set to maximum time
     self.total_locked = amount
-    log Fund(recipient, amount)
 
-    return True
+
+@external
+def seed(amount: uint256):
+    # probably should guard this one
```
Comment (milkyklim) [inline contracts/VestingEscrowSimple.vy:78]: ?

File: contracts/VestingEscrowFactory.vy [hunk 2]
```diff
@@ -63,17 +55,23 @@ def deploy_vesting_contract(
     @param vesting_start Epoch time when tokens begin to vest
     """
     assert cliff_length <= vesting_duration  # dev: incorrect vesting cliff
-    escrow: address = create_forwarder_to(self.target)
-    assert ERC20(token).transferFrom(msg.sender, self, amount)  # dev: funding failed
-    assert ERC20(token).approve(escrow, amount)  # dev: approve failed
-    VestingEscrowSimple(escrow).initialize(
+    assert vesting_duration > 0 # dev: duration must be > 0
+    escrow: address = create_from_blueprint(
```
Comment (banteg) [inline contracts/VestingEscrowFactory.vy:59]: let's discuss minimal proxy vs blueprint [trade-offs](https://docs.vyperlang.org/en/stable/built-in-functions.html#chain-interaction). how does the deployment cost change? are the savings on no delegatecall worth it given vesting contracts aren't getting called that often?

File: contracts/VestingEscrowSimple.vy [hunk 3]
```diff
@@ -35,28 +35,20 @@ cliff_length: public(uint256)
...
 @external
-def __init__():
-    # ensure that the original contract cannot be initialized
-    self.initialized = True
-
-
-@external
-@nonreentrant('lock')
-def initialize(
+def __init__(
```
Comment (banteg) [inline contracts/VestingEscrowSimple.vy:43]: it's cool we could use a proper `__init__` with blueprints.

File: contracts/VestingEscrowSimple.vy (hunk 1 above)
Comment (banteg) [reply contracts/VestingEscrowSimple.vy:78]: same as before, i'd rather store that a contract has been initialized and not allow repeating any actions that are only done once.

File: contracts/VestingEscrowSimple.vy [hunk 4]
```diff
@@ -133,7 +126,7 @@ def locked() -> uint256:
 
 
 @external
-def claim(beneficiary: address = msg.sender, amount: uint256 = MAX_UINT256):
+def claim(beneficiary: address = msg.sender, amount: uint256 = max_value(uint256)):
```
Comment (banteg) [inline contracts/VestingEscrowSimple.vy:129]: changes for release notes:
- add tether support
- anyone could claim on behalf or recipient

the latter should probably be a flag that you enable

File: contracts/VestingEscrowSimple.vy [hunk 5]
```diff
@@ -183,7 +176,7 @@ def apply_transfer_ownership():
     """
     assert msg.sender == self.future_admin  # dev: future admin only
     self.admin = msg.sender
-    self.future_admin = ZERO_ADDRESS
+    self.future_admin = empty(address)
```
Comment (banteg) [inline contracts/VestingEscrowSimple.vy:179]: i don't like this two-part commit thing in this contract, it's not a dao, maybe we should simplify it

File: contracts/VestingEscrowSimple.vy [hunk 6]
```diff
@@ -193,12 +186,14 @@ def renounce_ownership():
...
+    self.admin = empty(address)
+    log ApplyOwnership(empty(address))
+
 
 @external
-def collect_dust(token: address):
-    assert msg.sender == self.recipient  # dev: recipient only
-    assert (token != self.token.address or block.timestamp > self.disabled_at)
-    assert ERC20(token).transfer(self.recipient, ERC20(token).balanceOf(self))
+def collect_dust(token: address, beneficiary: address = msg.sender):
```
Comment (banteg) [inline contracts/VestingEscrowSimple.vy:195]: `token: address` can be changed to `token: ERC20`, then you won't have to wrap it in the interface every time

add dev comment

File: contracts/VestingEscrowSimple.vy [hunk 7]
```diff
@@ -193,12 +186,14 @@ def renounce_ownership():
...
+    recipient: address = self.recipient
+    assert msg.sender == recipient or recipient == beneficiary # dev: not authorized
+    assert (token != self.token.address or block.timestamp > self.disabled_at) # dev: can't collect
```
Comment (banteg) [inline contracts/VestingEscrowSimple.vy:198]: this is not correct behavior, if the vested token has been sent to the contract, the recipient should be able to claim the portion exceeding the remaining locked amount.

**Review by banteg — COMMENTED** [review]
looks very solid! thank you.

things to think about:
- should we simplify funding to one transfer?
- should we use blueprint vs minimal proxy?
- should rug pullability be defined at creation? counterpoint: you can bundle multiple actions when deploying from a dao or safe.
- should we change admin change to a single step?
- it should be possible to recover the vested token sent to the contract minus still locked amount, add a test demonstrating that.

File: contracts/VestingEscrowFactory.vy [hunk 8]
```diff
@@ -63,17 +55,23 @@ def deploy_vesting_contract(
...
         recipient,
         amount,
         vesting_start,
         vesting_start + vesting_duration,
         cliff_length,
+        salt=convert(msg.sender, bytes32),  # Ensures unique deployment per caller
+        code_offset=3,
     )
+
+    assert ERC20(token).transferFrom(msg.sender, self, amount, default_return_value=True) # dev: funding failed
```
[the remark this answers is not in the document]
Comment (banteg) [reply contracts/VestingEscrowFactory.vy:72]: decided to do this and implicitly assume the escrow created from the factory is funded

File: contracts/VestingEscrowFactory.vy (hunk 2 above)
Comment (banteg) [reply contracts/VestingEscrowFactory.vy:59]: decided to revert back to minimal proxies based on usage stats of previous versions.

File: contracts/VestingEscrowSimple.vy (hunk 5 above)
Comment (banteg) [reply contracts/VestingEscrowSimple.vy:186]: decided to delete the ability to change admin altogether, based on usage stats (this feature was only used in one escrow) and on the basis of vesting being an agreement between two parties. it makes no sense for one party to be able to swap for someone else.

File: contracts/VestingEscrowSimple.vy [hunk 9]
```diff
@@ -144,18 +145,24 @@ def claim(beneficiary: address = msg.sender, amount: uint256 = max_value(uint256
 
 
 @external
-def rug_pull():
+def rug_pull(ts: uint256 = block.timestamp):
```
Comment (banteg) [inline contracts/VestingEscrowSimple.vy:148]: should it be possible to set rugged funds recipient?

**Comment by banteg** [comment]
another decision:
- `open_claim` by default to enable smart contract recipients like 0xsplits.xyz
- `recipient` can call `set_open_claim(bool)` which can set `open_claim` to `False` to disable this behavior
- `set_open_claim` is a noop if the value doesn't change and emits `SetOpenClaim(bool)` if the value is changed
- the check in `claim()` is changed to `assert msg.sender == recipient or (self.open_claim and recipient == beneficiary)`

**Comment by banteg** [comment]
writing the full spec in #17

File: contracts/VestingEscrowSimple.vy [hunk 10]
```diff
@@ -151,39 +151,39 @@ def claim(beneficiary: address = msg.sender, amount: uint256 = max_value(uint256
...
 
     assert self.token.transfer(beneficiary, ruggable, default_return_value=True)
 
-    self.admin = empty(address)
+    self.owner = empty(address)
 
-    log SetFree(admin)
-    log VestingTerminated(self.recipient, admin, ruggable, ts)
+    log Disowned(owner)
+    log Revoked(self.recipient, owner, ruggable, ts)
```
Comment (banteg) [inline contracts/VestingEscrowSimple.vy:174]: we should also log beneficiary

File: contracts/VestingEscrowFactory.vy [hunk 11]
```diff
@@ -73,7 +73,7 @@ def deploy_vesting_contract(
     assert vesting_duration > 0  # dev: duration must be > 0
     assert recipient not in [self, empty(address), token, owner]  # dev: wrong recipient
```
Comment (banteg) [inline contracts/VestingEscrowFactory.vy:74]: let's relax to just empty(address)

**Pull request merged**